scheduler: fix quadratic performance with spread blocks

tgross opened the pull request:
When the scheduler picks a node for each evaluation, the
`LimitIterator` provides at most 2 eligible nodes for the
`MaxScoreIterator` to choose from. This keeps scheduling fast while
producing acceptable results because the results are binpacked.

Jobs with a `spread` block (or node affinity) remove this limit in
order to produce correct spread scoring. This means that every
allocation within a job with a `spread` block is evaluated against
_all_ eligible nodes. Operators of large clusters have reported that
jobs with `spread` blocks that are eligible on a large number of nodes
can take longer than the nack timeout to evaluate (60s). Typical
evaluations are processed in milliseconds.

In practice, it's not necessary to evaluate every eligible node for
every allocation on large clusters, because the `RandomIterator` at
the base of the scheduler stack produces enough variation in each pass
that the likelihood of an uneven spread is negligible. Note that
feasibility is checked before the limit, so this only impacts the
number of _eligible_ nodes available for scoring, not the total number
of nodes.

This changeset sets the iterator limit for "large" `spread` block and
node affinity jobs to be equal to the number of desired
allocations. This brings an example problematic job evaluation down
from ~3min to ~10s. The included tests ensure that we have acceptable 
spread results across a variety of large cluster topologies.

---

Note for reviewers: I have some additional PRs to make that'll improve our ability to debug this kind of problem in the future (ex. instrumenting the iterators, tooling for handling large snapshots), but I've intentionally kept them out of this PR so we have a clean bugfix for backporting. I'll push those PRs up over the next couple days.

schmichael commented on scheduler/spread_test.go:
```diff
@@ -568,3 +571,239 @@ func Test_evenSpreadScoreBoost(t *testing.T) {
+		return err
+	}
+	i := 0
+	nodesToRacks := map[string]string{}
+	for {
+		raw := iter.Next()
+		if raw == nil {
+			break
+		}
+		node := raw.(*structs.Node)
+		rack, ok := node.Meta["rack"]
+		if ok {
+			nodesToRacks[node.ID] = rack
+		}
+		i++
+	}
+
+	// Collapse the count of allocations per node into a list of
+	// counts. The results should be a single count or, if more than
+	// one count, they should be no more than one apart.
+	racksToAllocCount := map[string]int{}
```
Since `racksToAllocCount` does not track racks with `0` allocs, your assertion that all racks are within `1` doesn't cover some racks being 0 and some racks being 2.

With this change `TestSpreadOnLargeCluster/nodes=10k_many_uneven_racks_allocs=500` sometimes fails for me (`seed=1640040961673918997` is one example).

I think initializing all racks to 0 in racksToAllocCount makes the test more precise, however I think it's debatable whether the failed case is "good enough" or not.

```suggestion
	nodesToRacks := map[string]string{}
	racksToAllocCount := map[string]int{}
	for {
		raw := iter.Next()
		if raw == nil {
			break
		}
		node := raw.(*structs.Node)
		rack, ok := node.Meta["rack"]
		if ok {
			nodesToRacks[node.ID] = rack
			racksToAllocCount[rack] = 0
		}
		i++
	}

	// Collapse the count of allocations per node into a list of
	// counts. The results should be a single count or, if more than
	// one count, they should be no more than one apart.
```

tgross replied:
> I think initializing all racks to 0 in racksToAllocCount makes the test more precise, however I think it's debatable whether the failed case is "good enough" or not.

Augh, good catch... my initial tests never included enough tests to ever hit 0 on a rack and I added those 500 rack tests late. I'll look at how we can improve the tuning of both the value and the assertion; I definitely don't want us to have a flaky test here.

schmichael reviewed (approved):
Great work. I wonder if we should improve our documentation too. I think this paragraph could be expanded from:

> Spread criteria are treated as a soft preference by the Nomad scheduler. If no nodes match a given spread criteria, placement is still successful.

To something like:

> Spread criteria are treated as a soft preference by the Nomad scheduler. If no nodes match a given spread criteria, placement is still successful. To avoid scoring every node for every placement, allocations may not be perfectly spread. Spread works best on attributes with similar number of nodes: identically configured racks or similarly configured datacenters.

If anyone needs absolute precision with `spread` we could make the timeout for evaluations configurable. Given the high concurrency of scheduling workers, operators may tolerate a small number of very slow (eg 5 minute) evaluations.

tgross commented on scheduler/spread_test.go:
```diff
@@ -568,3 +571,239 @@ func Test_evenSpreadScoreBoost(t *testing.T) {
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
```
Good catch!

tgross commented:
I've picked up your suggestions and added the docs change. I've adjusted the test so that it's a bit less strict (effectively allowing for [0, 1, 2] which we need for the large rack counts) and produces more readable error outputs.

I've also added a note to the change in `stack.go` that notes that the limit value was empirically determined so that future developers will know there's nothing magical about 100 so they can feel free to adjust in the future if the scheduler changes significantly.
